### PR TITLE
Recycle pages used by outgoing publications

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today `PublicationTransportHandler.PublicationContext` allocates a bunch
of memory for serialized cluster states and diffs, but it uses a plain
`BytesStreamOutput` which means that the backing pages are allocated by
the `BigArrays#NON_RECYCLING_INSTANCE`. With this commit we pass in a
proper `BigArrays` so that the pages being used can be recycled.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
Hmm, the failure is a genuine problem: we simulate node reboots in `CoordinatorTests` and we block any further activity on the rebooted instance, but we don't release any memory they allocated.

**Comment by original-brownbear** [comment]
Thanks David, I always wanted this :)

> Hmm, the failure is a genuine problem: we simulate node reboots in CoordinatorTests and we block any further activity on the rebooted instance, but we don't release any memory they allocated.

I remember us running into this before when you tried this previously. Maybe it's ok to leave this as a TODO and just not use pooled buffers in tests that block all activity on a node like that? Not the nicest solution, but realistically if we want to block all activity on a node then that means no releasing of memory and we do have all kinds of tests for all kinds of failure scenarios/disconnects/you-name-it that provide coverage for this?

I suppose we could invest more effort and create another version of `BigArrays` that allows us some more fine grained control over this and selectively release all memory for a blocked node but that seems quite cumbersome?

**Comment by DaveCTurner** [comment]
It's actually not that bad it seems, see https://github.com/DaveCTurner/elasticsearch/commit/f5ce1551708c5029bd98ab12991af59c265563de#diff-0681cc9adce47fddd18bdfb39428f871bb4b592a40829fd6b58eeea7e1495358R1533-R1537 although that's not all, we also have to clean up messages that went to a blackhole (waiting a day works but makes the tests super-slow) and also there's another blackhole for messages that get sent to a node that has rebooted. Not sure what else I'll find yet, mostly I'm just working through failures where the cluster gets left in a broken state by the test.

**Comment by DaveCTurner** [comment]
failure is https://github.com/elastic/elasticsearch/issues/58946, @elasticmachine please run elasticsearch-ci/part-1

**Comment by DaveCTurner** [comment]
I've run through the `CoordinatorTests` over 100 times forcing the use of `MockBigArrays` without seeing any further leaks or other failures, so I believe this is ready to review now.

File: server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java [hunk 1]
```diff
@@ -314,96 +342,98 @@ public void onFailure(Exception e) {
...
 
         private void sendFullClusterState(DiscoveryNode destination, ActionListener<PublishWithJoinResponse> listener) {
-            BytesReference bytes = serializedStates.get(destination.getVersion());
+            assert refCount() > 0;
+            ReleasableBytesReference bytes = serializedStates.get(destination.getVersion());
             if (bytes == null) {
                 try {
                     bytes = serializeFullClusterState(newState, destination.getVersion());
-                    serializedStates.put(destination.getVersion(), bytes);
+                    final ReleasableBytesReference existingBytes = serializedStates.putIfAbsent(destination.getVersion(), bytes);
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java:359]: I wonder, can't we just do `computeIfAbsent` here instead and block on concurrent access? This is all on the generic pool anyways if there's contention isn't it? In any case no matter the thread this runs on, seems like blocking on the locked CHM key is still cheaper on the system overall than serializing twice? (it's an edge case either way but using `computeIfAbsent` is less code here I think? :)

**Review by original-brownbear — COMMENTED** [review]
Spent quite some time thinking this one through and couldn't find a leak I think. This looks really nice, thanks David!
 
Just one quick question on the concurrency handling (mainly to make sure I'm not missing any detail :)).

File: server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java:359]: Yeah that's a good point, I wasn't sure if `computeIfAbsent` ran the supplier twice sometimes but you made me go and read the docs and they say that it won't so it should be fine. Definitely a corner case of a corner case to be blocking here but I suppose we may as well not burn CPU while we wait given that we'll be waiting about as long either way. No risk of deadlock either.

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java]: Ok see 13bfe0c58214c22ec6df00e7d4fba7dc6138b017 (it triggered a few more changes/simplifications too)

File: server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTransportHandlerTests.java [hunk 2]
```diff
@@ -63,16 +86,213 @@ public ClusterState apply(ClusterState part) {
...
+                        }), responsesLatch::countDown));
+                    requestsLatch.countDown();
+                });
+            }
+
+            assertTrue(requestsLatch.await(10, TimeUnit.SECONDS));
+            context.decRef();
+            assertTrue(responsesLatch.await(10, TimeUnit.SECONDS));
+        } finally {
+            ThreadPool.terminate(threadPool, 10, TimeUnit.SECONDS);
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTransportHandlerTests.java]: ++ thanks, done in 7ca5255e6e075e9a0457874e22e87fd6ce5d51d2

**Review by original-brownbear — APPROVED** [review]
LGTM, if CI is happy I'm happy :) thanks David!

**Pull request merged**